src/probnum/filtsmooth/gaussfiltsmooth/kalman_utils.py

```diff
@@ -1,401 +1,419 @@
+#
+#     dim = len(A)
+#     zeros = np.zeros((dim, dim))
+#     blockmat = np.block(
+#         [
+#             [SC_past.T @ A.T, SC_past.T],
+#             [SQ.T, zeros],
+#             [zeros, SC_futu.T @ G.T],
+#         ]
+#     )
+#     big_triu = np.linalg.qr(blockmat, mode="r")
+#     SC = big_triu[dim : 2 * dim, dim:]
+#
+#     new_cov_cholesky = triu_to_positive_tril(SC)
+#     new_cov = new_cov_cholesky @ new_cov_cholesky.T
+#     new_mean = unsmoothed_rv.mean + G @ (smoothed_rv.mean - predicted_rv.mean)
+#     return pnrv.Normal(new_mean, new_cov, cov_cholesky=new_cov_cholesky), {}
+#
+#
+# ########################################################################################################################
+# # Helper functions
+# ########################################################################################################################
+#
+#
+# def linear_system_matrices(model, rv, start, stop, _linearise_at):
+#     """Extract the linear system matrices from Transition objects in order to apply
+#     square-root steps."""
+#     if isinstance(model, statespace.LTISDE):
+#         model = model.discretise(stop - start)
+#     elif isinstance(model, DiscreteEKFComponent):
+#         compute_jacobian_at = _linearise_at if _linearise_at is not None else rv
+#         model.linearize(compute_jacobian_at)
+#         model = model.linearized_model
+#     H = model.state_trans_mat_fun(start)
+#     SR = model.proc_noise_cov_cholesky_fun(start)
+#     shift = model.shift_vec_fun(start)
+#     return H, SR, shift
+#
+#
+# # used for predict() and measure(), but more general than that,
+# # so it has a more general name than the functions below.
+# def cholesky_update(
+#     S1: np.ndarray, S2: typing.Optional[np.ndarray] = None
+# ) -> np.ndarray:
+#     r"""Compute Cholesky update/factorization :math:`C C^\top = S_1 S_1^\top + S_2 S_2^\top`.
+#
+#     This can be used in various ways.
+#     For example, :math:`S_1` and :math:`S_2` do not need to be Cholesky factors; any matrix square-root is sufficient.
+#     As long as :math:`C C^\top = S_1 S_1^\top + S_2 S_2^\top` is well-defined (and admits a Cholesky-decomposition),
+#     :math:`S_1` and :math:`S_2` do not even have to be square.
+#     """
+#     # doc might need a bit more explanation in the future
+#     # perhaps some doctest or so?
+#     if S2 is not None:
+#         stacked_up = np.vstack((S1.T, S2.T))
+#     else:
+#         stacked_up = np.vstack(S1.T)
+#     upper_sqrtm = np.linalg.qr(stacked_up, mode="r")
+#     return triu_to_positive_tril(upper_sqrtm)
+#
+#
+# def triu_to_positive_tril(triu_mat: np.ndarray) -> np.ndarray:
+#     r"""Change an upper triangular matrix into a valid lower Cholesky factor.
+#
+#     Transpose, and change the sign of the diagonals to '+' if necessary.
+#     """
+#     tril_mat = triu_mat.T
+#     with_pos_diag = tril_mat @ np.diag(np.sign(np.diag(tril_mat)))
+#     return with_pos_diag
```

Comment:
The entire file should be removed before merging into master, imo.

Reply:
oooops